File: ImasLiveDB/Domain/UseCases/EditPermissionRules.swift
import Foundation

/// オープン編集 (作成/修正/削除・コミュニティ投稿) の権限判定。
///
/// 判定本体は imas-core (Rust) の `domain/edit_permission_rules.rs` にあり、
/// `EditPermissionRules` (Record: isSignedIn / isBanned の 2 射影) と
/// `EditTapOutcome` (present / promptLogin / ignore) の型自体も uniffi 生成
/// バインディングが提供する。ここは Swift らしい computed property の呼び口を
/// 生成関数へ委譲するだけの薄い拡張 (判定を書いたら負け)。
/// なぜ BAN を見るか・導線の出し分けの意図は edit_permission_rules.rs に記載。
///
/// 実際の認証状態 (`AuthService.shared`) から組み立てるのは `EditPermission` の役目。
extension EditPermissionRules {
    /// 編集 UI を実際に出してよいか (ログイン済み かつ 未 BAN)。
    var canEdit: Bool { editPermissionCanEdit(rules: self) }

    /// 未ログインで編集導線を押した時にログイン誘導を出すべきか。
    var shouldPromptLogin: Bool { editPermissionShouldPromptLogin(rules: self) }

    /// 編集 / 新規作成ボタン自体を表示してよいか (BAN 済みのみ隠す)。
    var showEditAffordance: Bool { editPermissionShowEditAffordance(rules: self) }

    /// 編集導線を押した時に何をするか。優先順は「編集可 > ログイン誘導 > 無視」。
    var outcomeOnEditTap: EditTapOutcome { editPermissionOutcomeOnEditTap(rules: self) }
}

File: imas-core/src/inbound/edit_permission_rules.rs
//! オープン編集権限判定の FFI 面。ロジックは domain::edit_permission_rules。
//!
//! `EditPermissionRules` は認証状態の 2 bool だけの射影 Record。UI の各判定
//! (ボタン表示可否・押下時の分岐など) がそれぞれ 1 回の FFI 呼び出しで完結する。

use crate::domain::edit_permission_rules::{EditPermissionRules, EditTapOutcome};

/// 編集 UI を実際に出してよいか (ログイン済み かつ 未 BAN)。
#[uniffi::export]
pub fn edit_permission_can_edit(rules: EditPermissionRules) -> bool {
    rules.can_edit()
}

/// 未ログインで編集導線を押した時にログイン誘導を出すべきか。
#[uniffi::export]
pub fn edit_permission_should_prompt_login(rules: EditPermissionRules) -> bool {
    rules.should_prompt_login()
}

/// 編集 / 新規作成ボタン自体を表示してよいか (BAN 済みのみ隠す)。
#[uniffi::export]
pub fn edit_permission_show_edit_affordance(rules: EditPermissionRules) -> bool {
    rules.show_edit_affordance()
}

/// 編集導線を押した時の分岐 (present / prompt_login / ignore)。
#[uniffi::export]
pub fn edit_permission_outcome_on_edit_tap(rules: EditPermissionRules) -> EditTapOutcome {
    rules.outcome_on_edit_tap()
}
